src/parser/ir/ir_parser_test.cc

```diff
@@ -93,6 +98,71 @@ INSTANTIATE_TEST_SUITE_P(IrParserTest, IrParserTest,
 }  // module m0
 )"));
 
+struct ParserInputAndExpectedOutput {
+  absl::string_view input;
+  absl::string_view output;
+};
+
+class IrParserNormalizingTest : public testing::TestWithParam<ParserInputAndExpectedOutput> {};
+
+TEST_P(IrParserNormalizingTest, DoublesAreNormalizedInTestOperations) {
+  auto program_text = GetParam();
+  IrProgramParser ir_parser;
+  auto result = ir_parser.ParseProgram(program_text.input);
+  EXPECT_EQ(IRPrinter::ToString(*result.module, *result.ssa_names),
+            program_text.output);
+}
+
+static const ParserInputAndExpectedOutput kNormalizedIR[] = {
```

Comment:
Is there a reason why we are using kNormalizedIR? Ideally we would want the input parsed text and outputs to be same. Also, the order of attributes seems to be different after normalization.

Reply:
Yes :)

Round trip tests are useful, but they only check that normalized output can be parsed.
In my experience at least, parsers should not restrict input unnecessarily.
So, we should allow users to input non-normalized input (to avoid user frustration) and output normalized input.

This is what these tests are for (because it can be hard to work out what a normalized double will be, and similar with the order of attributes).

Reply:
I see. Sounds good!